### PR TITLE
API

### PR DESCRIPTION
**Pull request by Alesia-V175**
1. Task: https://github.com/rolling-scopes-school/tasks/tree/master/react/modules/module04
2. Screenshot :
image
image
image

3. Deploy: https://alesia-v157-react-rsschool.netlify.app/
4. Done 09.04.2023 / Deadline 10.04.2023
5. Score: 15 / 15

*if you have warnings about a cookie in a cross-site request (SameSite attribute), please ignore it ! 
It's server site problem and I can't do anything from my side. I hope for your understanding*

There is you can find full PR to main brach: 
https://github.com/Alesia-V175/react-rs-school/pull/38

Instructions:

You need to clone this repo and checkout the branch 'api' branch or 'develop' branch.

1. npm install (install required packages)
2. npm run dev (start project in developer mode)
3. npm run lint (check code style)
4. npm run build (prepare application build)
5. npm test (run tests)

If you have any questions, please connect with me in Discord #Alesia (Alesia-V175)#6265

### CODE REVIEW

File: src/components/CardItem/CardItem.tsx [hunk 1]
```diff
@@ -1,12 +1,29 @@
...
 const CardItem = (card: ICardItem): JSX.Element => {
+  const [cardDetailed, setCardDetailed] = useState<ICardItem | undefined>();
+
   const [date] = new Date(card.created_at).toISOString().split('T');
 
+  const handleClick = () => {
+    if (!cardDetailed) {
+      Api.getCard(card.id)
+        .then((cardItem) => {
+          setCardDetailed(cardItem);
```
Comment (Ksarelto) [inline src/components/CardItem/CardItem.tsx:17]: use async await:
```
const handleClick = async () => {
    if (!cardDetailed) {
      const response = await Api.getCard(card.id)
      setCardDetailed(response);
    }
  };
```

File: src/components/CardItem/CardItem.tsx [hunk 2]
```diff
@@ -1,12 +1,29 @@
...
+        });
+    }
+  };
+
   return (
-    <div className={styles.card__wrap}>
+    <div className={styles.card__wrap} onClick={handleClick}>
+      {cardDetailed ? (
+        <FullCardItem card={cardDetailed} close={() => setCardDetailed(undefined)}/>
+      ) : (<></>)}
```
Comment (Ksarelto) [inline src/components/CardItem/CardItem.tsx:26]: try to not use ternary operator in render:
```
{
   cardDetailed &&
   (
        <FullCardItem card={cardDetailed} close={() => setCardDetailed(undefined)}/>
   )
}
```

File: src/components/FullCardItem/FullCardItem.test.tsx [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { render, screen } from '@testing-library/react';
+import { vitest } from 'vitest';
+import FullCardItem from './FullCardItem';
+import { ICardItem } from '../../types/interfaces';
+import database from '../../database/database.json';
+
+const data = vitest.fn();
+const card: ICardItem = database[0];
+beforeEach(() => {
```
Comment (Ksarelto) [inline src/components/FullCardItem/FullCardItem.test.tsx:9]: empty line before 'beforeEach'

and 'beforeEach' usually inside 'describe'

File: src/components/FullCardItem/FullCardItem.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+          {card.tags_preview && card.tags_preview.length ? (
+            <div>Tags:
+              {card.tags_preview && card.tags_preview.map((value) => (
+                <span
+                  className={styles.card__description_detailed}
+                  key={value.title}> {value.title},
+                </span>
+              ))}
+            </div>
+          ) : (<></>)}
```
Comment (Ksarelto) [inline src/components/FullCardItem/FullCardItem.tsx:44]: no need in ternary operator)):
```
{
  card.tags_preview &&
  !!card.tags_preview.length && 
  (
     <div>Tags:
         {card.tags_preview && card.tags_preview.map((value) => (
            <span
                className={styles.card__description_detailed}
                key={value.title}> {value.title},
             </span>
          ))}
      </div>
   ) 
}
```

File: src/components/Search/Search.tsx [hunk 5]
```diff
@@ -1,16 +1,24 @@
...
+const Search = ({ searchCards }: ISearch): JSX.Element => {
   const [searchValue, setSearchValue] = useState(() => {
-    const savedSearchValue = localStorage.getItem(KeyLocalStorage.search);
+    let savedSearchValue = localStorage.getItem(KeyLocalStorage.search);
 
-    if (savedSearchValue) {
-      return savedSearchValue;
+    if (!savedSearchValue) {
+      savedSearchValue = '';
     }
```
Comment (Ksarelto) [inline src/components/Search/Search.tsx:19]: can u set as:
```
const [searchValue, setSearchValue] = useState(() => localStorage.getItem(KeyLocalStorage.search) ?? '';
```
?

File: src/components/Search/Search.tsx [hunk 6]
```diff
@@ -20,12 +28,14 @@ const Search = (): JSX.Element => {
...
-    const savedSearchValue = localStorage.getItem(KeyLocalStorage.search);
+    let savedSearchValue = localStorage.getItem(KeyLocalStorage.search);
 
-    if (savedSearchValue) {
-      setSearchValue(savedSearchValue);
+    if (!savedSearchValue) {
+      savedSearchValue = '';
     }
 
+    setSearchValue(savedSearchValue);
```
Comment (Ksarelto) [inline src/components/Search/Search.tsx:37]: almosat the same as above

File: src/components/Search/Search.tsx [hunk 7]
```diff
@@ -35,6 +45,12 @@ const Search = (): JSX.Element => {
     setSearchValue(event.target.value);
   };
 
+  const handleKeyEvent = (event: React.KeyboardEvent<HTMLElement>): void => {
+    if (event.key === 'Enter') {
```
Comment (Ksarelto) [inline src/components/Search/Search.tsx:49]: for handling Enter you can wrap your inputs in form => its onSubmit event by default watch for Enter keypress

File: src/Api/Api.ts [hunk 8]
```diff
@@ -1,12 +1,13 @@
 import { API } from '../constants/apiConstants';
 import { RequestMethod } from '../types/enums';
+import { ICardItem, ISearchParams } from '../types/interfaces';
 
 class Api {
   private static API_URL = API.API_URL;
 
   private static ACCESS_KEY = API.ACCESS_KEY;
```
[the remark this answers is not in the document]
Comment (Alesia-V175) [reply src/Api/Api.ts:8]: fixed

File: src/Api/Api.ts [hunk 9]
```diff
@@ -23,6 +24,42 @@ class Api {
...
+  }
+
+  public static async getCard(id: string): Promise<ICardItem | undefined> {
+    const url = `${this.API_URL}/photos/${id}`;
+
+    const response = await fetch(url, {
+      method: RequestMethod.GET,
+      headers: {
+        Authorization: `Client-ID ${this.ACCESS_KEY}`,
+      },
```
[the remark this answers is not in the document]
Comment (Alesia-V175) [reply src/Api/Api.ts:53]: created separate function apiRequest

File: src/components/CardItem/CardItem.tsx (hunk 2 above)
Comment (Alesia-V175) [reply src/components/CardItem/CardItem.tsx:26]: fixed

File: src/components/FullCardItem/FullCardItem.tsx (hunk 4 above)
Comment (Alesia-V175) [reply src/components/FullCardItem/FullCardItem.tsx:44]: deleted  ternary operator)

File: src/components/Search/Search.tsx (hunk 7 above)
Comment (Alesia-V175) [reply src/components/Search/Search.tsx:49]: added form wrapper

**Pull request merged**